app/src/main/java/io/appium/settings/RecorderService.java

```diff
@@ -0,0 +1,135 @@
+    public int onStartCommand(final Intent intent, final int flags, final int startId) {
+        Log.v(TAG, "onStartCommand:intent=" + intent);
+
+        int result = START_STICKY;
+        final String action = intent != null ? intent.getAction() : null;
+        Log.v(TAG, "onStartCommand:action string: " + action);
+        if (ACTION_RECORDING_START.equals(action)) {
+            showNotification(); // TODO is this really necessary
+
+            MediaProjectionManager mMediaProjectionManager =
+                    (MediaProjectionManager) getSystemService(Context.MEDIA_PROJECTION_SERVICE);
+
+            if (mMediaProjectionManager != null) {
+                startScreenRecord(mMediaProjectionManager, intent);
+            } else {
+                Log.e(TAG, "mMediaProjectionManager null");
+            }
+        } else if (ACTION_RECORDING_STOP.equals(action)) {
+            Log.v(TAG, "onStartCommand:intent=" + "stop");
+            stopScreenRecord();
+            result = START_NOT_STICKY;
+        }
+
+        return result;
+    }
+
+    /**
+     * start screen recording
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void startScreenRecord(MediaProjectionManager mediaProjectionManager,
+                                   final Intent intent) {
+        if (recorderThread == null) {
+            Log.v(TAG, "recorder thread null");
+
+            final int resultCode = intent.getIntExtra(ACTION_RECORDING_RESULT_CODE, 0);
+            // get MediaProjection
+            final MediaProjection projection = mediaProjectionManager.getMediaProjection(resultCode,
+                    intent);
+            if (projection != null) {
+                Log.v(TAG, "projection different from null");
+                final DisplayMetrics metrics = getResources().getDisplayMetrics();
+                int rawWidth = metrics.widthPixels;
+                int rawHeight = metrics.heightPixels;
+                boolean isRotated = false;
+                if (rawWidth > rawHeight) {
+                    // landscape mode
+                    isRotated = true;
+                    /* TODO we need to rotate frames that comes from virtual screen before writing to file via muxer,
+                    *  for handling landscape mode properly, rotateYuvImages somehow fast and reliable
+                    *  for now, just flip width and height to fake and force portrait mode instead of landscape
+                    * */
+                    rawWidth = metrics.heightPixels;
+                    rawHeight = metrics.widthPixels;
+                }
+                Log.v(TAG, String.format("startRecording:(%d,%d)(%d,%d)",
+                        metrics.widthPixels, metrics.heightPixels, rawWidth, rawHeight));
+                String outputFilePath = intent.getStringExtra(ACTION_RECORDING_FILENAME);
+                if (outputFilePath != null) {
+                    recorderThread = new RecorderThread(projection, outputFilePath,
```

Comment:
should we also validate if the file name does not contain invalid characters, which cannot be used in linux file names?

Reply:
AFAIK, there is no "file path validation" method on java (at least on java 8) but i may be wrong, there are other type of validation techniques (like creating empty file for checking) but they seems a bit hacky solutions, i am open to ideas ? 👍 
 thanks

Reply:
I tend to use simple regex validation like in https://stackoverflow.com/questions/2679699/what-characters-allowed-in-file-names-on-android

Reply:
okay done https://github.com/sirmordred/io.appium.settings/commit/dc8dc277448fb3e401559698c10ab17ca8030193 i searched a bit on the net and there were lots of regex that some of them were very restrictive and some does not catch all forbidden chars, so i kanged android's internal validation from aosp :) https://android.googlesource.com/platform/frameworks/base/+/refs/heads/master/core/java/android/os/FileUtils.java#995